### DISCUSSION TITLE
mypy doesn't grok the new lifespan context manager

### DISCUSSION THREAD

**Discussion by kthy** [fastapi/fastapi#9241] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from collections.abc import AsyncGenerator
from contextlib import asynccontextmanager

from fastapi import FastAPI


@asynccontextmanager
async def lifespan(_: FastAPI) -> AsyncGenerator[None, None]:
    ...
    yield
    ...


app = FastAPI(lifespan=lifespan)
```

### Description

I have taken the [example from the docs of the lifespan async context manager](https://fastapi.tiangolo.com/advanced/events/#lifespan) - new in 0.93 - shaved it down to the bare minimum, and added typing hints.

When running this code through mypy-1.1.1 it complains about the type of the `lifespan()` context manager:

```bash
$ pipenv run mypy lifespan_example.py
Loading .env environment variables...
lifespan_example.py:14: error:
  Argument "lifespan" to "FastAPI" has incompatible type 
    "Callable[[FastAPI], _AsyncGeneratorContextManager[None]]";
  expected
    "Union[Callable[[Starlette], AbstractAsyncContextManager[None]], Callable[[Starlette], AbstractAsyncContextManager[Mapping[str, Any]]], None]"  [arg-type]
Found 1 error in 1 file (checked 1 source file)
```

I am unsure whether this is an issue with FastAPI, mypy, contextlib, or me. 🤷 

### Operating System

Linux

### Operating System Details

Ubuntu 20.04 running in WSL2.

### FastAPI Version

0.94

### Python Version

3.10.8

**Comment by Kludex** [comment]
@adriangb 😬

**Reply by adriangb** [reply]
https://github.com/encode/starlette/pull/2077

cc @tiangolo so we can coordinate next release.

@kthy for now I would just add a `# type: ignore`, sorry for the inconvenience

**Comment by Kludex** [comment, marked as answer]
As @adriangb said, this is an issue on Starlette's side. We'll be handling it over there.

Thanks for the report. 🙏

**Reply by tiangolo** [reply]
Just adding here, thank you @kthy for the really good report. :bow: 

> shaved it down to the **bare minimum**

This is key, greatly appreciated, and in most cases, it's already half of the work. Thanks! :coffee: :cake:
